src/metabase/legacy_mbql/normalize.cljc

```diff
@@ -1050,57 +1050,54 @@
 ;;; |                                             REMOVING EMPTY CLAUSES                                             |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
-(declare remove-empty-clauses)
-
-(defn- remove-empty-clauses-in-map [m path]
-  (let [m (into (empty m) (for [[k v] m
-                                :let  [v (remove-empty-clauses v (conj path k))]
-                                :when (some? v)]
-                            [k v]))]
-    (when (seq m)
-      m)))
-
-(defn- remove-empty-clauses-in-sequence* [xs path]
-  (let [xs (mapv #(remove-empty-clauses % (conj path ::sequence))
-                 xs)]
+(declare remove-empty-clauses path->special-remove-empty-clauses-fn)
+
+(defn- remove-empty-clauses-in-map [m special-fns]
+  (not-empty (reduce-kv (fn [m k v]
+                          (let [v' (remove-empty-clauses v (get special-fns k))]
+                            (cond (nil? v') (dissoc m k)
+                                  (identical? v v') m
+                                  :else (assoc m k v'))))
+                        m m)))
+
+(defn- remove-empty-clauses-in-sequence* [xs special-fns]
+  (let [special-fns (::sequence special-fns)
+        xs (#?(:clj metabase.util.performance/mapv :default mapv) #(remove-empty-clauses % special-fns) xs)]
     (when (some some? xs)
       xs)))
 
 (defmulti ^:private remove-empty-clauses-in-mbql-clause
-  {:arglists '([clause path])}
-  (fn [[tag] _path]
+  {:arglists '([clause special-fns])}
+  (fn [[tag] _special-fns]
     tag))
 
 (defmethod remove-empty-clauses-in-mbql-clause :default
-  [clause path]
-  (remove-empty-clauses-in-sequence* clause path))
+  [clause special-fns]
+  (remove-empty-clauses-in-sequence* clause special-fns))
 
 (defmethod remove-empty-clauses-in-mbql-clause :offset
-  [[_tag opts expr n] path]
-  [:offset opts (remove-empty-clauses expr (conj path :offset)) n])
+  [[_tag opts expr n] special-fns]
+  [:offset opts (remove-empty-clauses expr (:offset special-fns)) n])
 
-(defn- remove-empty-clauses-in-sequence [x path]
+(defn- remove-empty-clauses-in-sequence [x special-fns]
   (if (mbql-clause? x)
-    (remove-empty-clauses-in-mbql-clause x path)
-    (remove-empty-clauses-in-sequence* x path)))
+    (remove-empty-clauses-in-mbql-clause x special-fns)
+    (remove-empty-clauses-in-sequence* x special-fns)))
 
 (defn- remove-empty-clauses-in-join [join]
-  (remove-empty-clauses join [:query]))
+  (remove-empty-clauses join (:query path->special-remove-empty-clauses-fn)))
 
 (defn- remove-empty-clauses-in-source-query [{native? :native, :as source-query}]
   (if native?
-    (-> source-query
-        (set/rename-keys {:native :query})
-        (remove-empty-clauses [:native])
-        (set/rename-keys {:query :native}))
```

Comment:
I don't get this part at all. It renames `:native` key to `:query`, then calls `remove-empty-clauses` with the path being `[:native]`. If you go into `path->special-remove-empty-clauses-fn` you'll see that `:native` has a special handling function that is `identity`. So it does nothing. Finally, we rename `:query` back to `:native`.

So is this just a big no-op or do I miss something?